Write Mixins/Integration guide (#1362)
huggingface/huggingface_hub: docs/source/guides/integrations.mdx

# Integrate any ML framework with the Hub

> The Huggingface Hub makes it easy to host and share models with the community. It already has support for

Reviewer: ```suggestion
The Huggingface Hub makes hosting and sharing models with the community easy. It supports
```

Author: ```suggestion
The Hugging Face Hub makes hosting and sharing models with the community easy. It supports
```
(applying both suggestions :D )

[dozens of libraries](https://huggingface.co/docs/hub/models-libraries) in the Open Source ecosystem. We are always

> working on expanding this support to push Machine Learning forward. The `huggingface_hub` library plays a key role in

Reviewer: ```suggestion
working on expanding this support to push collaborative Machine Learning forward. The `huggingface_hub` library plays a key role in
```

this process, allowing any Python script to easily push and load files.

> When talking about integrating a library to the Hub, we have 4 main topics:

Reviewer: ```suggestion
When talking about integrating a library to the Hub, we have four main topics:
```

> When talking about integrating a library to the Hub, we have 4 main topics:

Reviewer: Note we also have https://huggingface.co/docs/hub/models-adding-libraries

> When talking about integrating a library to the Hub, we have 4 main topics:

Reviewer 2: Maybe it would be nice to mention the first two are essential as I think inference support is often cherry on top.

> 1. **Push to Hub:** implement a `push_to_hub()`-like method to upload a model to the Hub.

Reviewer 2: I feel like directly mentioning `push_to_hub` and `from_pretrained` in this point of the guide introduces question marks to user, I'd rather just say what it does here but it's my opinion :D @stevhliu @Wauplin

Reviewer 3: That's a great point! It does kind of assume you're familiar with the `push_to_hub` method. You could either describe it as Merve suggested, or add a link to the `push_to_hub` docstring :)

Reviewer 2: there's already explanations on them as you go down in the guide, it's just I felt people might be intimidated not knowing what it is at the beginning written like this :)

>    This includes the model weights as well as the modelcard and any other relevant information/data necessary to run the model (example: logs from training).

Reviewer: ```suggestion
   This includes the model weights, the modelcard and any other relevant information/data necessary to run the model (example: logs from training).
```

>    This includes the model weights as well as the modelcard and any other relevant information/data necessary to run the model (example: logs from training).

Reviewer 2: ```suggestion
   This includes the model weights, as well as [the model card](https://huggingface.co/docs/huggingface_hub/how-to-model-cards) and any other relevant information or data necessary to run the model (for example, training logs).
```
This is completely optional but I thought it kinda assumes too much that the person knows all about the model card (and I feel like it's the second most important asset after weights)

2. **Download from Hub:** implement a `from_pretrained`-like method to load a model from the Hub.

>    Takes care of downloading the model configuration/weights and instantiating it.

Reviewer 2: ```suggestion
   The method should download the model configuration/weights and load the model.
```

> 3. **Inference API:** collaborate with the HF Team to make it possible to run inference on our servers, using your library.

Reviewer: ```suggestion
3. **Inference API:** run inference of models supported by your library for free on our servers.
```

4. **Widgets:** collaborate with the HF Team to display a widget on the landing page of your models on the Hub. It allows users to quickly try a model from the browser.

> In this guide, we will focus on topic 1 and 2. If you are interested in Inference and Widgets, please let us know!

Reviewer: ```suggestion
In this guide, we will focus on the first two topics. If you are interested in Inference and Widgets, please let us know!
```

You can also reach out to us if you are integrating your library with the Hub and want to be listed [in our docs](https://huggingface.co/docs/hub/models-libraries).

## A bit of theory: use class inheritance

> As we saw above, we are interested in having 2 methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).

Reviewer: ```suggestion
As we saw above, we are interested in having two methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).
```

> As we saw above, we are interested in having 2 methods: one to upload files (`push_to_hub`) and one to download files (`from_pretrained`).

Reviewer 3: ```suggestion
As we saw above, there are two main methods to include in your library to integrate it with the Hub: upload files (`push_to_hub`) and download files (`from_pretrained`).
```

In most cases, a library already implements its model using a Python class. The class contains the properties of the model

> and methods to load, run, train, evaluate,... it. It would be nice to extend the class with the 2 new methods!

Reviewer: ```suggestion
and methods to load, run, train, and evaluate it. It would be nice to extend the class with the two new methods!
```

> and methods to load, run, train, evaluate,... it. It would be nice to extend the class with the 2 new methods!

Reviewer 3: ```suggestion
and methods to load, run, train, evaluate,... it. You can extend the class to include functionality for uploading and downloading files to the Hub!
```

Author: ```suggestion
and methods to load, run, train, and evaluate it. You can extend the class to include functionality for uploading and downloading files to the Hub!
```

> The recommended approach to do so is to use inheritance, and more precisely mixins. A [Mixin](https://stackoverflow.com/a/547714)

Reviewer: Do we want to make this the official recommended approach? This is more complex/scary than the usually push to hub/download_from_hub with the HTTP endpoints and requires having classes in their libraries

> The recommended approach to do so is to use inheritance, and more precisely mixins. A [Mixin](https://stackoverflow.com/a/547714)

Reviewer 3: You can swap out best for whatever type of approach you want to make this, whether its recommended or something else.

```suggestion
The best way to do this is by using inheritance, specifically mixins. A [Mixin](https://stackoverflow.com/a/547714)
```

is a class that is meant to extend an existing class with a set of specific features using multiple inheritance.

> `huggingface_hub` defined its own mixin [`ModelHubMixin`]. The key here is to understand its behavior and how to customize it.

Reviewer 3: ```suggestion
`huggingface_hub` provides its own mixin, the [`ModelHubMixin`]. The key here is to understand its behavior and how to customize it.
```

> The [`ModelHubMixin`] class defines 3 *public* methods (`push_to_hub`, `save_pretrained` and `from_pretrained`) and 2

Author: Will do! I'm still mentioning the public methods to emphasize how it will be beneficial to users.

   [the model card](https://huggingface.co/docs/huggingface_hub/how-to-model-cards) and any other relevant information
   or data necessary to run the model (for example, training logs). This method is often called `push_to_hub()`.
2. **Download from Hub:** implement a method to load a model from the Hub. The method should download the model
   configuration/weights and load the model. This method is often called `from_pretrained` or `load_from_hub()`.
3. **Inference API:** use our servers to run inference on models supported by your library for free.
4. **Widgets:** display a widget on the landing page of your models on the Hub. It allows users to quickly try a model
   from the browser.

In this guide, we will focus on the first two topics. If you are interested in Inference and Widgets, you can follow
[this guide](https://huggingface.co/docs/hub/models-adding-libraries#set-up-the-inference-api). In both cases, you can
reach out to us if you are integrating your library with the Hub and want to be listed [in our docs](https://huggingface.co/docs/hub/models-libraries). 

> ## A first approach: helpers

Reviewer 3: A bit nitpicky, but I'd consider naming this section something like "A flexible approach: helpers"?. The "first approach" suggests users should try this first, which makes it seem like it's the recommended way to integrate (unless this is the recommended way 😛).

Author: Yep, good point :+1:

The first approach to integrate a library to the Hub is to actually implement the `push_to_hub` and `from_pretrained`
methods by yourself. This gives you full flexibility on which files you need to upload/download and how to handle inputs

> specific to your framework. You can refers to the two [upload files](./upload) and [download files](./download) guides

Reviewer 3: ```suggestion
specific to your framework. You can refer to the two [upload files](./upload) and [download files](./download) guides
```

> to learn more on how to do that. This is for example how FastAI integration is implemented (see [`push_to_hub_fastai`]

Reviewer 3: ```suggestion
to learn more about how to do that. This is, for example how the FastAI integration is implemented (see [`push_to_hub_fastai`]
```

and [`from_pretrained_fastai`]).

> Implementation can defer between libraries but the workflow is often similar.

Reviewer 3: ```suggestion
Implementation can differ between libraries, but the workflow is often similar.
```

### from_pretrained

This is how a `from_pretrained` method usually look like:

```python
def from_pretrained(model_id: str):
   # Download model from Hub
   cached_model = hf_hub_download(
      repo_id=repo_id,
      filename="model.pkl",
      library_name="fastai",
      library_version=get_fastai_version(),
   )

   # Load model
    return load_model(cached_model)
```

### push_to_hub

> The `push_to_hub` often require a bit more complexity to handle repo creation, generate the model card and save weights.

Reviewer 3: ```suggestion
The `push_to_hub` method often requires a bit more complexity to handle repo creation, generate the model card and save weights.
```

A common approach is to save all of these files in a temporary folder, upload it and then delete it.

```python
def push_to_hub(model, repo_name: str) -> None:
   api = HfApi()

   # Create repo if not existing yet and get the associated repo_id
   repo_id = api.create_repo(repo_name, exist_ok=True)

   # Save all files in a temporary directory and push them in a single commit
   with TemporaryDirectory() as tmpdir:
      tmpdir = Path(tmpdir)

      # Save weights
      save_model(model, tmpdir / "model.safetensors")

      # Generate model card
      card = generate_model_card(model)
      (tmpdir / "README.md").write_text(card)

      # Save logs
      # Save figures
      # Save evaluation metrics
      # ...

      # Push to hub
      return api.upload_folder(repo_id=repo_id, folder_path=tmpdir)
```

This is of course only an example. If you are interested in more complex manipulations (delete remote files, upload

> weights on the fly, persist weights locally,...) please refer to the [upload files](./upload) guide.

Reviewer 3: ```suggestion
weights on the fly, persist weights locally, etc.) please refer to the [upload files](./upload) guide.
```

> ### Limitations

Reviewer 3: I'd move the Limitations section above the `from_pretrained` method. I think it'd be nice to have a full picture of the pros/cons of using this approach before actually trying to implement it.

Author: I understand your point but I think it could be confusing to users -at least with the text I wrote-. If I move the "limitations" section up, the reader would not really have in mind the method signatures/implementation that I already list everything that is missing from it.

What I can do is to add a sentence in the introduction to say "hey, we will see 2 different approaches, each of them have advantages/drawbacks so choose what's best for you". So that when the user reads the first approach he/she already have in mind that there are some limitations (without knowing which ones yet). What do you think?

Reviewer 3: Great idea! 👍

> While being flexible, this approach has some drawbacks, especially in term of maintenance. Huggingface users are often
> used to additional features when working with `huggingface_hub`. For example, when load files from the Hub, it is
> common to offer parameters like

Reviewer 3: ```suggestion
While being flexible, this approach has some drawbacks, especially in terms of maintenance. Hugging Face users are often
used to additional features when working with `huggingface_hub`. For example, when loading files from the Hub, it is
common to offer parameters like:
```

- `token`: to download from a private repo
- `revision`: to download from a specific branch
- `cache_dir`: to cache files in a specific directory
- `force_download`/`resume_download`/`local_files_only`: to reuse the cache or not
- `api_endpoint`/`proxies`: configure HTTP session

When pushing models, similar parameters are supported:
- `commit_message`: custom commit message
- `private`: create a private repo if missing
- `create_pr`: create a PR instead of pushing to `main`
- `branch`: push to a branch instead of the `main` branch
- `allow_patterns`/`ignore_patterns`: filter which files to upload
- `token`
- `api_endpoint`
- ...

All of these parameters can be added to the implementations we saw above and passed to the `huggingface_hub` methods.
However, if a parameter changes or a new feature is added, you will need to update your package. Supporting those
parameters also means more documentation to maintain on your side. To see how to mitigate these limitations, let's jump
to our next section **class inheritance**.

## A more complex approach: class inheritance

> ### A bit of theory

Reviewer 3: I don't think we need a header here

            revision=revision,
            cache_dir=cache_dir,
            force_download=force_download,
            proxies=proxies,
            resume_download=resume_download,
            token=token,
            local_files_only=local_files_only,
         )

      # Load model and return - custom logic depending on your framework
      model = cls(**model_kwargs)
      state_dict = torch.load(model_file, map_location=torch.device(map_location))
      model.load_state_dict(state_dict, strict=strict)
      model.eval()
      return model
```

And that's it! Your library now enables users to upload and download files to and from the Hub.

## Quick comparison

Let's quickly sum up the two approaches we saw with their advantages and drawbacks. The table below is only indicative.
Your framework might have some specificities that you need to address. This guide is only here to give guidelines and
ideas on how to handle integration. In any case, feel free to contact us if you have any questions!

<!-- Generated using https://www.tablesgenerator.com/markdown_tables -->
| Integration | Using helpers | Using [`ModelHubMixin`] |
|:---:|:---:|:---:|

> | User experience | Good.<br>`model = load_from_hub(...)`<br>`push_to_hub(model, ...)` | Good.<br>`model = MyModel.from_pretrained(...)`<br>`model.push_to_hub(...)` |

Reviewer 3: Maybe remove "Good" here since they're both good and it doesn't really tell you much; might be better to only compare usage.

| Flexibility | Very flexible.<br>You fully control the implementation. | Less flexible.<br>Your framework must have a model class. |

> | Maintenance | More maintenance to add support for configuration, new features,... Might also requires to fix issues reported by users. | Less maintenance as most of the interactions with the Hub are implemented in `huggingface_hub`. |

Reviewer 3: ```suggestion
| Maintenance | More maintenance to add support for configuration, and new features. Might also require fixing issues reported by users. | Less maintenance as most of the interactions with the Hub are implemented in `huggingface_hub`. |
```
